From 8568bc56865c8aa1aec33789ec04811ba1b50e61 Mon Sep 17 00:00:00 2001
From: Viorel Suman <viorel.suman@nxp.com>
Date: Thu, 16 Aug 2018 14:40:34 +0300
Subject: [PATCH 4423/5242] MLK-19154-4: ASoC: imx_spdif: set SPDIF ROOT clock
 freq as function of rate

commit  007808c2a9c3e15a399def820d90b5c1ea896a5c from
https://source.codeaurora.org/external/imx/linux-imx.git

Set SPDIF master clock frequency as function of rate.

Signed-off-by: Viorel Suman <viorel.suman@nxp.com>
(cherry picked from commit 407430a03994e4acff508afe8c9772680558c1c5)
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 sound/soc/fsl/imx-spdif.c |   27 +++++++++++++++++++++++++++
 1 file changed, 27 insertions(+)

diff --git a/sound/soc/fsl/imx-spdif.c b/sound/soc/fsl/imx-spdif.c
index 7994243..440f7cd 100644
--- a/sound/soc/fsl/imx-spdif.c
+++ b/sound/soc/fsl/imx-spdif.c
@@ -12,12 +12,38 @@
 #include <linux/module.h>
 #include <linux/of_platform.h>
 #include <sound/soc.h>
+#include "fsl_spdif.h"
 
 struct imx_spdif_data {
 	struct snd_soc_dai_link dai;
 	struct snd_soc_card card;
 };
 
+#define CLK_8K_FREQ    24576000
+#define CLK_11K_FREQ   22579200
+
+static int imx_spdif_hw_params(struct snd_pcm_substream *substream,
+		struct snd_pcm_hw_params *params)
+{
+	struct snd_soc_pcm_runtime *rtd = substream->private_data;
+	struct device *dev = rtd->card->dev;
+	int ret;
+	u64 rate = params_rate(params);
+	unsigned int freq;
+
+	freq = do_div(rate, 8000) ? CLK_11K_FREQ : CLK_8K_FREQ;
+	ret = snd_soc_dai_set_sysclk(rtd->cpu_dai, STC_TXCLK_SPDIF_ROOT,
+		freq, SND_SOC_CLOCK_OUT);
+	if (ret)
+		dev_err(dev, "failed to set cpu sysclk: %d\n", ret);
+
+	return ret;
+}
+
+static struct snd_soc_ops imx_spdif_ops = {
+	.hw_params = imx_spdif_hw_params,
+};
+
 static int imx_spdif_audio_probe(struct platform_device *pdev)
 {
 	struct device_node *spdif_np, *np = pdev->dev.of_node;
@@ -45,6 +71,7 @@ static int imx_spdif_audio_probe(struct platform_device *pdev)
 	data->dai.platform_of_node = spdif_np;
 	data->dai.playback_only = true;
 	data->dai.capture_only = true;
+	data->dai.ops = &imx_spdif_ops;
 
 	if (of_property_read_bool(np, "spdif-out"))
 		data->dai.capture_only = false;
-- 
1.7.9.5

